File: TeamCode/src/main/java/com/info1robotics/bobot/tasks/DigitalTask.java
package com.info1robotics.bobot.tasks;

import android.drm.DrmStore;
import android.util.Pair;

import com.info1robotics.bobot.Common.GamepadEx;
import com.info1robotics.bobot.OpModes.templates.TeleOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class DigitalTask extends Task {
    public String button;
    public int gamepad;

    public static class ActionPair {
        Type type;
        Task task;

        public ActionPair(Type type, Task task) {
            this.type = type;
            this.task = task;
        }
    }

    public enum Type {
        PRESS,
        HOLD,
        RELEASE
    }

    public ArrayList<ActionPair> tasks = new ArrayList<>();

    public DigitalTask(String button, int gamepad, ActionPair ...children) {
        this.button = button;
        this.gamepad = gamepad;
        tasks.addAll(Arrays.asList(children));
    }

    @Override
    public void tick() {
        TeleOpMode teleOp = (TeleOpMode) context;
        tasks.forEach(pair -> {
            if (pair.task.isRunning()) pair.task.tick();
            GamepadEx gamepadEx = (gamepad == 1) ? teleOp.gamepadEx : teleOp.gamepadEx2;
            switch (pair.type) {
                case PRESS:
                    if (gamepadEx.getButtonDown(button)) {
                        pair.task.start(context);
                    }
                    break;
                case RELEASE:
                    if (gamepadEx.getButtonUp(button)) {
                        pair.task.start(context);
                    }
                    break;
                case HOLD:
                    if (gamepadEx.getButton(button)) {
                        pair.task.start(context);
                    }
                    break;
            }
        });
    }
}

File: TeamCode/src/main/kotlin/com/info1robotics/bobot/tasks/DigitalTask.kt
package com.info1robotics.bobot.tasks

import com.info1robotics.bobot.Common.GamepadEx
import com.info1robotics.bobot.opmodes.TeleOpMode

/**
 * Listener for digital buttons.
 * Never ends, intended to be used with [AllTasks].
 * See [TaskBuilder.digital].
 * @param button The digital button to listen for.
 */
class DigitalTask(private val button: GamepadEx.Digital, private val gamepad: Int) : Task() {
    enum class Type {
        PRESS,
        RELEASE,
        HOLD
    }
    var tasks = mutableMapOf<Type, SyncTasks>()

    override fun tick() {
//        println("ticking")
        if (context !is TeleOpMode) {
            throw Exception("DigitalTask can only be used in TeleOp")
        }
        val teleOpMode = (context) as TeleOpMode
        tasks.forEach { (type, action) ->
            if (action.isRunning()) action.tick()
            val gamepadEx = if (gamepad == 1) teleOpMode.gamepadEx else teleOpMode.gamepadEx2
//            println(gamepadEx)
            when (type) {
                Type.PRESS -> {
                    if (gamepadEx.getButtonDown(button)) {
                        action.start(context)
                    }
                }
                Type.RELEASE -> {
                    if (gamepadEx.getButtonUp(button)) {
                        action.start(context)
                    }
                }
                Type.HOLD -> {
                    if (gamepadEx.getButton(button)) {
                        println("Holding $button")
                        action.start(context)
                    }
                }
            }
        }
    }

    /**
     * Registers tasks to be synchronously executed when the button is executed with the given type.
     * @param type The action type. See [Type].
     * @param init Lambda to initialise the action. See [SyncTasks].
     */
    fun on(type: Type, init: SyncTasks.() -> Unit): SyncTasks {
        val task = SyncTasks()
        task.init()
        tasks[type] = task
        return task
    }
}